### PR TITLE
[SofaCore] Add ObjectFactory_test.cpp

### PR DESCRIPTION
**Pull request by damienmarchal**
So we can validate how the ObjectFactory should behave. 

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        .add< TestObject<long> >();
+int B3 = RegisterObject("Loads a plugin and import its content into the current namespace. ")
+        .add< TestObject2<int> >();
+
+class ObjectFactory_test: public BaseTest
+{
+public:
+    void testDuplicatedRegistration()
+    {
+        {
```
Comment (guparan) [inline SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp]: Why the double brackets?

File: SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    void testValidAlias()
+    {
+        ASSERT_TRUE(ObjectFactory::getInstance()->addAlias("FirstAlias", "TestObject"));
+    }
+
+    void testInvalidAlias()
+    {
+        {
```
Comment (guparan) [inline SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp]: Why the double brackets?

File: SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+};
+
+int A = RegisterObject("Loads a plugin and import its content into the current namespace. ")
+        .add< TestObject<int> >();
+int B1 = RegisterObject("Loads a plugin and import its content into the current namespace. ")
+        .add< TestObject<double> >();
+int B2 = RegisterObject("Loads a plugin and import its content into the current namespace. ")
+        .add< TestObject<long> >();
+int B3 = RegisterObject("Loads a plugin and import its content into the current namespace. ")
+        .add< TestObject2<int> >();
```
Comment (guparan) [inline SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp]:
```suggestion
int A = RegisterObject("Dummy test object.")
        .add< TestObject<int> >();
int B1 = RegisterObject("Dummy test object.")
        .add< TestObject<double> >();
int B2 = RegisterObject("Dummy test object.")
        .add< TestObject<long> >();
int B3 = RegisterObject("Dummy test object.")
        .add< TestObject2<int> >();
```

File: SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+class ObjectFactory_test: public BaseTest
+{
+public:
+    void testDuplicatedRegistration()
+    {
+        {
+            EXPECT_MSG_EMIT(Warning);
+            int C = RegisterObject("Loads a plugin and import its content into the current namespace. ")
```
Comment (guparan) [inline SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp]:
```suggestion
            int C = RegisterObject("Already registered object.")
```

File: SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp (hunk 2 above)
Comment (damienmarchal) [reply SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp]: Because of EXPECT_MSG_EMIT. The validity of the test is scoped by the brackets.

**Pull request merged**